Dependencies/xr/Source/ARCore/XR.cpp

```diff
@@ -860,6 +862,69 @@ namespace xr
             }
         }
 
+        void UpdateFeaturePointCloud()
+        {
+            if (!IsTracking() || !FeaturePointCloudEnabled)
+            {
+                return;
+            }
+
+            // Get the feature point cloud from ArCore.
+            ArPointCloud *pointCloud = nullptr;
+            int32_t numberOfPoints = 0;
+            const int32_t* pointCloudIDs = nullptr;
+            const float *pointCloudData = nullptr;
+            ArStatus status = ArFrame_acquirePointCloud(session, frame, &pointCloud);
+
+            if (status != AR_SUCCESS)
+            {
+                FeaturePointCloud.clear();
+                return;
+            }
+
+            try
+            {
+                ArPointCloud_getNumberOfPoints(session, pointCloud, &numberOfPoints);
+                ArPointCloud_getData(session, pointCloud, &pointCloudData);
+                ArPointCloud_getPointIds(session, pointCloud, &pointCloudIDs);
+
+                FeaturePointCloud.resize(numberOfPoints);
+                for (int32_t i = 0; i < numberOfPoints; i++)
+                {
+                    FeaturePointCloud.emplace_back();
+                    auto& featurePoint = FeaturePointCloud.back();
+
+                    int32_t dataIndex = i * 4;
+                    featurePoint.x = pointCloudData[dataIndex];
+                    featurePoint.y = pointCloudData[dataIndex + 1];
+                    featurePoint.z = -1 * pointCloudData[dataIndex + 2];
+                    featurePoint.confidenceValue = pointCloudData[dataIndex + 3];
+
+                    // Check to see if this point ID exists in our point cloud mapping if not add it to the map.
+                    const int32_t id = pointCloudIDs[i];
+                    auto featurePointIterator = featurePointIDMap.find(id);
+                    if (featurePointIterator != featurePointIDMap.end())
+                    {
+                        featurePoint.ID = featurePointIterator->second;
+                    }
+                    else
+                    {
+                        featurePoint.ID = nextFeaturePointID++;
+                        featurePointIDMap.insert({id, featurePoint.ID});
+                    }
+                }
+            }
+            catch (std::exception)
+            {
+                // Release the point cloud to free its memory.
+                ArPointCloud_release(pointCloud);
+                throw;
+            }
+
+            // Release the point cloud to free its memory.
+            ArPointCloud_release(pointCloud);
```

Comment:
Can't we just use a finally?
```suggestion
            finally
            {
                // Release the point cloud to free its memory.
                ArPointCloud_release(pointCloud);
            }
```

Reply:
No finally in C++ because well designed APIs use RAII, but the ARCore code was designed for C.